FreeTubeApp/FreeTube
# Update subscription view to be able to load videos from video cache per channel

## Pull Request Type


## Related issue



A variant (?) of #3665

## Description

#3665 restores the "original" behaviour (well the code that was there but disabled somehow)
This is a different video caching strategy
Videos (and potentially other info like live streams and shorts) are cached per **channel** instead of profile
The difference is that every profile can potentially just load from cache if all the channels of selected profile have required content in cache, without the limitation of only caching one profile at a time

## Screenshots 

Similar to 3665 one so no repeating here

## Testing 




Preparation: 
- Create following profiles (other than default)
  - Profile A with **most** but not all channels
  - Profile B with only some channels from A
  - Profile C with some channels from A (1 is enough), some from the channel(s) **excluded** in A
  - Profile D with no channel

**Before starting each test please reopen window/app/refresh to ensure local state is reset**

A. Auto fetching disabled
A1. Loading from cache after all channel cached
- Switch to default profile (if not already)
- Ensure no video loaded
- Load videos from remote (via button, KB shortcut whatever)
- Ensure videos loaded
- Switch to profile A
- Ensure **correct** videos loaded, no network request made for fetching videos
- Switch to profile B
- Ensure **correct** videos loaded, no network request made for fetching videos
- Switch to profile C
- Ensure **correct** (no) videos loaded, no network request made for fetching videos
- Switch to default profile
- Ensure **correct** (all) videos loaded, no network request made for fetching videos

A2a. Refresh subscriptions effects - With default profile loaded first
- Switch to default profile (if not already)
- Load videos from remote
- Ensure videos loaded
- Switch to profile A
- Load videos from remote
- Switch to default profile
- Ensure **all** videos loaded, no network request made for fetching videos
  (due to only cache for channels in profile A refreshed, but cache for not refreshed channel are not purged)
- Switch to profile A
- Ensure **correct** videos loaded, no network request made for fetching videos
- Switch to profile B
- Ensure **correct** videos loaded, no network request made for fetching videos

A2b. Refresh subscriptions effects - Without default profile loaded first
- Switch to default profile (if not already)
- **DO NOT** load videos from remote
- Switch to profile A
- Load videos from remote
- Switch to default profile
- Ensure **NO** videos loaded
  (due to not all channels are cached)
- Switch to profile A
- Ensure **correct** videos loaded, no network request made for fetching videos
- Switch to profile B
- Ensure **correct** videos loaded, no network request made for fetching videos
  (due to `Profile B with only some channels from A`)

A3. After subscribing new channel(s)
- Switch to default profile (if not already)
- Load videos from remote
- Ensure videos loaded
- Visit a random not subscribed channel, subscribe it
- Go back to subscriptions view
- Ensure **NO** videos loaded (due to not all subscribed channel cached)

B. Auto fetching enabled
B1. Loading from `subscriptionsCacheForAllSubscriptionsProfile`
- Same as A1 except `Load videos from remote` is done automatically

B2a. Refresh subscriptions effects - With default profile loaded first
- Switch to default profile (if not already)
- Ensure videos loaded
- Switch to profile A
- Load videos from remote
- Switch to default profile
- Ensure **all** videos loaded, no network request made for fetching videos
  (due to only cache for channels in profile A refreshed, but cache for not refreshed channel are not purged)
- Switch to profile A
- Ensure **correct** videos loaded, no network request made for fetching videos
- Switch to profile B
- Ensure **correct** videos loaded, no network request made for fetching videos

B2b. Refresh subscriptions effects - Without default profile loaded first
- Switch to non subscription view
- Switch to profile A
- Ensure videos loaded
- Switch to default profile
- Ensure videos loaded, with no. of requests matching no. of channels
  (i.e. all videos are fetched from remote not from cache)
- Switch to profile A
- Ensure **correct** videos loaded, no network request made for fetching videos
- Switch to profile B
- Ensure **correct** videos loaded, no network request made for fetching videos
  (due to `Profile B with only some channels from A`)

B3. After subscribing new channel(s)
- Switch to default profile (if not already)
- Ensure videos loaded
- Visit a random not subscribed channel, subscribe it
- Go back to subscriptions view
- Ensure videos loaded, with no. of requests matching no. of channels
  (i.e. all videos are fetched from remote not from cache)

## Desktop

- **OS:**
- **OS Version:**
- **FreeTube version:**

## Additional context

`Fetch Feed Automatically` is enabled by default
To avoid expectation issue when enabled all videos are fetch from remote (if cache for any channel absent)
